node/src/coordinator.rs

```diff
@@ -317,29 +342,155 @@ impl Coordinator {
+                tracing::info!("Key generation complete; Follower calls vote_pk.");
+                chain_txn_sender_cloned
+                    .send(ChainSendTransactionRequest::VotePk(ChainVotePkArgs {
+                        key_event_id: contract_event.id,
+                        public_key: affine_point_to_public_key(res.public_key)?,
+                    }))
+                    .await?;
+                //Ok(())
+                //});
+            }
+        } else {
+            let n_participants = mpc_config.participants.participants.len();
+            'leader: loop {
+                let mut contract_event = key_event_receiver.borrow_and_update().clone();
+                if !contract_event.started {
+                    tracing::info!(
+                        "Leader is starting ecdsa secp256k1 key generation for key id {:?}",
+                        contract_event.id
+                    );
+                    // open the channel
+                    let channel = network_client.new_channel_for_task(
+                        EcdsaTaskId::KeyGeneration {
+                            key_event: contract_event.id.clone(),
+                        },
+                        network_client.all_participant_ids(),
+                    )?;
+                    tracing::info!(
+                        "opened channel, sending start keygen on id: {:?}",
+                        channel.task_id()
+                    );
+                    chain_txn_sender
+                        .send(ChainSendTransactionRequest::StartKeygen(
+                            ChainStartKeygenArgs {},
+                        ))
+                        .await?;
+                    tracing::info!("sent start keygen, starting computation");
+                    // note: we could inline below function, but maybe better to wait and see how other
+                    // singauter schemes will be handled.
+                    let res = EcdsaSignatureProvider::run_key_generation_client(
+                        mpc_config.participants.threshold as usize,
+                        channel,
+                    )
+                    .await;
+                    let res = match res {
+                        Ok(res) => res,
+                        Err(e) => anyhow::bail!("error: {}", e),
+                    };
+                    tracing::info!("leader concluded computation, storing keyshare.");
+                    let keyshare = Keyshare {
+                        key_id: contract_event.id.clone(),
+                        data: KeyshareData::Secp256k1(res.clone()),
+                    };
+                    keyshare_storage.store_key(keyshare).await?;
+                    tracing::info!("leader stored keyshare.");
+                    let my_public_key = affine_point_to_public_key(res.public_key)?;
+                    let key_event_id = contract_event.id;
+                    while contract_event.completed.len() != n_participants - 1 {
```

Comment:
There's no need to wait for follower votes.

When leader finishes generation it can just go ahead and vote, and just enter the next iteration. Then it naturally waits for the contract to not be started again (or this state gets killed because we're in Running now) and it'll start another computation. We implemented all the great logic at the contract side so we don't need this extra coordination in the node :)